File: seguecodePlugin/seguecodePlugin.m
//
//  seguecodePlugin.m
//  seguecodePlugin
//
//  Created by Ian on 11/4/14.
//  Copyright (c) 2014 Adorkable. All rights reserved.
//

#import "seguecodePlugin.h"

#import "NSObject+Utility.h"
#import "NSNotification+Utility.h"
#import "NSURL+Utility.h"

#import "NSMutableDictionary+RunConfig.h"

static seguecodePlugin *sharedPlugin;

@interface seguecodePlugin()

@property (nonatomic, strong, readwrite) NSBundle *bundle;

@property (copy) NSURL *currentlyEditingStoryboardFileName;

@property (weak, readwrite) NSMenuItem *storyboardEnabled;

@end

@implementation seguecodePlugin

+ (void)pluginDidLoad:(NSBundle *)plugin
{
    static dispatch_once_t onceToken;
    NSString *currentApplicationName = [[NSBundle mainBundle] infoDictionary][@"CFBundleName"];
    if ([currentApplicationName isEqual:@"Xcode"]) {
        dispatch_once(&onceToken, ^{
            sharedPlugin = [[self alloc] initWithBundle:plugin];
        });
    }
}

+ (instancetype)sharedPlugin
{
    return sharedPlugin;
}

- (id)initWithBundle:(NSBundle *)plugin
{
    if (self = [super init] )
    {
        self.bundle = plugin;
        
        if ( [self pathToSegueCode] )
        {
            NSLog(@"seguecode found at %@", [self pathToSegueCode] );
            
            [self setupNotifications];
            
            [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                [self setupMenu];
            }];
        } else
        {
            NSLog(@"seguecode not found!");
        }
    }
    return self;
}

- (void)setupNotifications
{
    [ [NSNotificationCenter defaultCenter] addObserver:self
                                              selector:@selector(transitionFromOneFileToAnother:)
                                                  name:[NSNotification transitionFromOneFileToAnotherName]
                                                object:nil];
    [ [NSNotificationCenter defaultCenter] addObserver:self
                                              selector:@selector(ideEditorDocumentDidSave:)
                                                  name:[NSNotification ideEditorDocumentDidSaveName]
                                                object:nil];
/*
    [ [NSNotificationCenter defaultCenter] addObserver:self
                                              selector:@selector(logNotification:)
                                                  name:nil
                                                object:nil];
*/
}

- (void)transitionFromOneFileToAnother:(NSNotification *)notification
{
    if ( [notification isTransitionFromOneFileToAnother] )
    {
        NSURL *nextURL = [notification transitionFromOneFileToAnotherNextDocumentURL];
        NSRange begin = [nextURL rangeOfString:@"file://"];
        
        if (begin.location != NSNotFound && begin.length > 0)
        {
            if ( [ [nextURL pathExtension] isEqualToString:@"storyboard"] )
            {
                self.currentlyEditingStoryboardFileName = nextURL;
            } else
            {
                self.currentlyEditingStoryboardFileName = nil;
                NSLog(@"Ignoring %@", [nextURL pathExtension] );
            }
        } else
        {
            NSLog(@"Unable to find file:// in %@", nextURL);
            self.currentlyEditingStoryboardFileName = nil;
        }
        
        [self updateStoryboardEnabled];
    } else
    {
        NSLog(@"Unexpected notification %@ sent to %@",
              notification,
              NSStringFromSelector(@selector(transitionFromOneFileToAnother:) )
              );
    }
}

- (void)ideEditorDocumentDidSave:(NSNotification *)notification
{
    if ( [notification isIDEEditorDocumentDidSave] )
    {
        id notificationObject = [notification object];
        
        Class IBStoryboardDocument = NSClassFromString(@"IBStoryboardDocument");
        if ( [notificationObject isKindOfClass:IBStoryboardDocument] && self.currentlyEditingStoryboardFileName )
        {
            NSMutableDictionary *runConfig = [NSMutableDictionary runConfigForStoryboardAtURL:self.currentlyEditingStoryboardFileName];
            
            if (runConfig)
            {
                NSLog(@"Applying to %@", self.currentlyEditingStoryboardFileName);
                [self applyToStoryboardAtURL:self.currentlyEditingStoryboardFileName
                                withRunConfig:runConfig];
            } else
            {
                NSLog(@"Skipping %@, not configured for usage", self.currentlyEditingStoryboardFileName);
            }
        }
    }
}

-(void)logNotification:(NSNotification *)notification
{
    if ( [notification isXcodeEvent] )
    {
        NSLog(@"Notification: %@ %@", [notification name], [notification object]);
    }
}

- (BOOL)createDefaultRunConfigForStoryboardAtPath:(NSURL *)storyboardURL
{
    NSMutableDictionary *result = [NSMutableDictionary dictionary];
    result.combine = NO;
    result.outputPath = [NSString stringWithFormat:@"./Generated"];
    return [result writeRunConfigForStoryboardAtURL:storyboardURL];
}

- (BOOL)applyToStoryboardAtURL:(NSURL *)storyboardURL withRunConfig:(NSMutableDictionary *)runConfig
{
    NSPipe *pipe = [NSPipe pipe];
    
    NSTask *task = [[NSTask alloc] init];
    task.launchPath = [ [self pathToSegueCode] path];

    NSString *outputFolder = [NSString stringWithFormat:@"\"%@/%@\"", [ [storyboardURL URLByDeletingLastPathComponent] path], runConfig.outputPath];
    
    NSMutableArray *arguments = [NSMutableArray array];
    [arguments addObjectsFromArray:@[@"--outputPath", outputFolder] ];
    if (runConfig.combine)
    {
        [arguments addObject:@"--combine"];
    }
    if (runConfig.projectName)
    {
        [arguments addObjectsFromArray:@[@"--projectName", runConfig.projectName] ];
    }
    [arguments addObject:[NSString stringWithFormat:@"\"%@\"", [storyboardURL path] ] ];
    task.arguments = arguments;
    
    task.currentDirectoryPath = [ [storyboardURL URLByDeletingLastPathComponent] path];
    task.standardOutput = pipe;
    [task launch];
    [task waitUntilExit];
    
    NSData *outputData = [ [pipe fileHandleForReading] availableData];
    NSString *outputResult = [ [NSString alloc] initWithData:outputData encoding:NSUTF8StringEncoding];
    if (outputResult.length > 0)
    {
        NSLog(@"seguecode result:\n%@", outputResult);
    }

    return task.terminationStatus != 0;
}

- (NSURL *)pathToSegueCode
{
    return [self.bundle URLForResource:@"seguecode.bundle/Contents/MacOS/seguecode"
                         withExtension:nil];
}

- (NSMenuItem *)xcodeEditMenu
{
    NSMenuItem *result = [ [NSApp mainMenu] itemWithTitle:@"Edit"];
    if (result == nil)
    {
        NSLog(@"Unable to find Edit menu");
    }
    return result;
}

- (void)setupMenu
{
    NSMenuItem *editMenuItem = [self xcodeEditMenu];
    if (editMenuItem)
    {
        [ [editMenuItem submenu] addItem:[NSMenuItem separatorItem] ];
        
        NSMenuItem *storyboardEnabled = [ [NSMenuItem alloc] initWithTitle:@"Enable seguecode"
                                                                    action:@selector(toggleEnabled:)
                                                             keyEquivalent:@""];
        [storyboardEnabled setTarget:self];
        [ [editMenuItem submenu] addItem:storyboardEnabled];
        self.storyboardEnabled = storyboardEnabled;
    } else
    {
        NSLog(@"Cannot add seguecode menu items");
    }
}

- (void)toggleEnabled:(id)sender
{
    if (self.currentlyEditingStoryboardFileName)
    {
        if (![self enabledForStoryboardAtPath:self.currentlyEditingStoryboardFileName] )
        {
            [self createDefaultRunConfigForStoryboardAtPath:self.currentlyEditingStoryboardFileName];
        } else
        {
            [NSMutableDictionary removeRunConfigForStoryboardAtURL:self.currentlyEditingStoryboardFileName];
        }
        [self updateStoryboardEnabled];
    }
}

- (BOOL)enabledForStoryboardAtPath:(NSURL *)storyboardURL
{
    // TODO: why aren't we using the passed in value?
    return self.currentlyEditingStoryboardFileName && [NSMutableDictionary runConfigForStoryboardAtURL:self.currentlyEditingStoryboardFileName];
}

- (BOOL)validateMenuItem:(NSMenuItem *)menuItem
{
    BOOL result = YES;
    if (menuItem == self.storyboardEnabled)
    {
        result = self.currentlyEditingStoryboardFileName != nil;
    }
    return result;
}

- (void)updateStoryboardEnabled
{
    if (self.currentlyEditingStoryboardFileName)
    {
        [self.storyboardEnabled setEnabled:YES];
        if ( [self enabledForStoryboardAtPath:self.currentlyEditingStoryboardFileName] )
        {
            [self.storyboardEnabled setState:NSOnState];
        } else
        {
            [self.storyboardEnabled setState:NSOffState];
        }
    } else
    {
        [self.storyboardEnabled setEnabled:NO];
        [self.storyboardEnabled setState:NSOffState];
    }
}



- (void)dealloc
{
    [ [NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

File: seguecodePlugin/seguecodePlugin.swift
//
//  seguecodePlugin.swift
//  seguecode
//
//  Created by Ian Grossberg on 11/17/15.
//  Copyright © 2015 Adorkable. All rights reserved.
//

import AppKit

import CocoaLumberjack

import seguecodeKit

class seguecodePlugin: NSObject {

    private let bundle: NSBundle
    
    private var currentlyEditingStoryboardUrl : NSURL? = nil
    
    private var storyboardEnabledMenuItem : NSMenuItem? = nil
    
    init(bundle : NSBundle) {
        self.bundle = bundle
        
        super.init()
        
        self.setupNotifications()
        self.setupMenu()
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "applicationFinishedLoading:", name: NSApplicationDidFinishLaunchingNotification, object: nil)
    }
    
    func applicationFinishedLoading(notification : NSNotification) {
        
        NSNotificationCenter.defaultCenter().removeObserver(self, name: NSApplicationDidFinishLaunchingNotification, object: nil)
    }
    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    func setupNotifications() {
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "transitionFromOneFileToAnother:", name: NSNotification.TransitionFromOneFileToAnotherName, object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "ideEditorDocumentDidSave:", name: NSNotification.IdeEditorDocumentDidSaveName, object: nil)
  
        // For Notification Logging
        #if DEBUG
//        NSNotificationCenter.defaultCenter().addObserver(self, selector: "logNotification:", name: nil, object: nil)
        #endif
    }
    
    func setupMenu() {
        
        // TODO: GCD?
        NSOperationQueue.mainQueue().addOperationWithBlock { () -> Void in

            guard let submenu = Constants.Xcode.IDE.submenuInstance?.submenu else {
                DDLog.error("Unable to find submenu to insert seguecode options into")
                
                // TODO: kill plugin
                return
            }
            
            submenu.addItem(NSMenuItem.separatorItem())
            
            let storyboardEnabled = NSMenuItem(title: "Enable seguecode", action: "toggleEnabled:", keyEquivalent: "")
            self.storyboardEnabledMenuItem = storyboardEnabled
            storyboardEnabled.target = self
            
            submenu.addItem(storyboardEnabled)
            
            DDLog.debug("\(storyboardEnabled.title) menu item added to submenu")
        }
    }
    
    func toggleEnabled(sender : AnyObject) {

        guard let currentlyEditingStoryboardUrl = self.currentlyEditingStoryboardUrl else {
            return
        }

        do
        {
            if self.currentlyEditingStoryboardEnabled {
                try self.removeRunConfigForStoryboard(storyboardUrl: currentlyEditingStoryboardUrl)
            } else {
                try self.createDefaultRunConfigForStoryboard(storyboardUrl: currentlyEditingStoryboardUrl)
            }
        } catch let error as NSError {
            DDLog.error("While toggling Storyboard enabled file for Storyboard \(currentlyEditingStoryboardUrl): \(error)")
        }
    }
    
    private var currentlyEditingStoryboardEnabled : Bool {
        
        guard let currentlyEditingStoryboardUrl = self.currentlyEditingStoryboardUrl else {
            return false
        }
        
        do
        {
            return try RunConfig(forStoryboardAtUrl: currentlyEditingStoryboardUrl) != nil
        } catch let error as NSError {
            
            DDLog.error("While creating a RunConfig for Storyboard at Url \(currentlyEditingStoryboardUrl): \(error)")
            return false
        }
    }
}

extension seguecodePlugin {
    
/*    @objc
    override func validateMenuItem(menuItem: NSMenuItem) -> Bool {
        
        guard menuItem == self.storyboardEnabledMenuItem else {
            return super.validateMenuItem(menuItem)
        }
        
        return self.currentlyEditingStoryboardUrl != nil
    }*/
}

extension seguecodePlugin {

    func transitionFromOneFileToAnother(notification : NSNotification) {
        
        guard let nextUrl = notification.transitionFromOneFileToAnotherNextDocumentUrl else {
            
            // TODO:
            return
        }
        
//        if nextUrl.isFileReferenceURL() {
            
            if nextUrl.pathExtension == Constants.Storyboard.extensionString {
                
                self.currentlyEditingStoryboardUrl = nextUrl
            } else {
                
                // TODO:
            }
/*        } else {
            
            // TODO:
        }*/
        
        // TODO: only if currentlyEditingStoryboardUrl changed (not storyboard -> not storyboard should not trigger update)
        self.updateStoryboardEnabled()
    }
    
    func updateStoryboardEnabled() {

        guard self.currentlyEditingStoryboardUrl != nil else {
            
            self.storyboardEnabledMenuItem?.enabled = false
            self.storyboardEnabledMenuItem?.state = NSOffState
            return
        }
        
        self.storyboardEnabledMenuItem?.enabled = true
        if self.currentlyEditingStoryboardEnabled {
            self.storyboardEnabledMenuItem?.state = NSOnState
        } else {
            self.storyboardEnabledMenuItem?.state = NSOffState
        }
    }
    
    func createDefaultRunConfigForStoryboard(storyboardUrl storyboardUrl : NSURL) throws {
        let runConfig = RunConfig()
        try runConfig.writeForStoryboard(storyboardUrl: storyboardUrl)
    }
    
    func removeRunConfigForStoryboard(storyboardUrl storyboardUrl : NSURL) throws {
        try RunConfig.removeForStoryboard(storyboardUrl: storyboardUrl)
    }
    
    func ideEditorDocumentDidSave(notification : NSNotification) {

        guard let object = notification.object else {
            // TODO:
            return
        }
        
        guard let IBStoryboardDocumentClass = NSClassFromString("IBStoryboardDocument") else {
            // TODO:
            return
        }
        
        // TODO: Is this necessary if we have the next check?
        guard object.isKindOfClass(IBStoryboardDocumentClass) else {
            return
        }

        // TODO: check that object is the currently editing storyboard file
        guard let currentlyEditingStoryboardUrl = self.currentlyEditingStoryboardUrl else {
            return
        }
        
        do
        {
            guard let runConfig = try RunConfig(forStoryboardAtUrl: currentlyEditingStoryboardUrl) else {
                return
            }
            
            self.applyToStoryboard(currentlyEditingStoryboardUrl, runConfig: runConfig)
            
        } catch let error as NSError {
            NSLog("While opening Run Config for Storyboard \(currentlyEditingStoryboardUrl): \(error)")
        }
    }
    
    func applyToStoryboard(storyboardUrl : NSURL, runConfig : RunConfig) -> Bool {

        guard let outputUrl = NSURL(string: runConfig.outputPath, relativeToURL: storyboardUrl) else {
            
            DDLog.error("Output Path \"\(runConfig.outputPath)\" in Run Config is an invalid relative URL for storyboard file \"\(storyboardUrl)\"")
            return false
        }

        seguecode.parse(storyboardUrl, outputPath: outputUrl, projectName: runConfig.projectName, exportTogether: runConfig.combine, verboseLogging: false)
        return true
    }
    
    func logXcodeNotification(notification : NSNotification) {
        if notification.isXcodeEvent {
            NSLog("Notification: \(notification.name), \(notification.object)")
        }
    }
}
